src/macaron/slsa_analyzer/provenance/witness/__init__.py

```diff
@@ -120,41 +121,38 @@ def extract_repo_url(witness_payload: InTotoPayload) -> str | None:
     return None
 
 
-def extract_witness_provenance_subjects(witness_payload: InTotoPayload) -> set[WitnessProvenanceSubject]:
-    """Read the ``"subjects"`` field of the provenance to obtain the hash digests of each subject.
+def extract_build_artifacts_from_witness_subjects(witness_payload: InTotoPayload) -> list[InTotoV01Subject]:
+    """Extract subjects that are build artifacts from the ``"subject"`` field of the provenance.
+
+    Each artifact subject is assumed to have a sha256 digest. If a sha256 digest is not present for
+    a subject, that subject is ignored.
 
     Parameters
     ----------
     witness_payload : InTotoPayload
         The witness provenance payload.
-    extensions : list[str]
-        The allowed extensions of the subjects.
-        All subjects with names not ending in these extensions are ignored.
 
     Returns
     -------
-    dict[str, str]
-        A dictionary in which each key is a subject name and each value is the corresponding SHA256 digest.
+    list[InTotoV01Subject]
+        A list subjects in the ``"subject"`` field of the provenance that are build artifacts.
     """
-    if isinstance(witness_payload, InTotoV01Payload):
-        subjects = witness_payload.statement["subject"]
-        subject_digests = set()
-
-        for subject in subjects:
-            name = subject["name"]
-            digest = subject["digest"]
-
-            sha256 = digest.get("sha256")
-            if not sha256 or not isinstance(sha256, str):
-                continue
-
-            subject_digests.add(
-                WitnessProvenanceSubject(
-                    subject_name=name,
-                    sha256_digest=sha256,
-                )
-            )
-
-        return subject_digests
-
-    return set()
+    if not isinstance(witness_payload, InTotoV01Payload):
+        return []
+
+    subjects = witness_payload.statement["subject"]
+    artifact_subjects = []
+    for subject in subjects:
+        # Filter all subjects attested by the product attestor, which records all changed and
+        # created files in the build process.
+        # Documentation: https://github.com/in-toto/witness/blob/main/docs/attestors/product.md
+        if not subject["name"].startswith("https://witness.dev/attestations/product/v0.1/file:"):
```

Comment:
This can match files that are not eventually published artifacts because these are files recorded by the Witness product attestors, which include all files created by the build.
However, this is sufficient for our internal use case that we are trying to support.